draft-specifications/CapTP Specification.md

```diff
@@ -0,0 +1,978 @@
+promise is created for the value of the response. They can be chained together
+in what is called [Promise Pipelining](#promise-pipelining), whereby messages
+are sent to the promise which should be delivered if the promise resolves.
+
+## [Promise and Resolver Objects](#promise-objects)
+
+Promises work like regular objects in CapTP. Promises come as a pair, there is a
+"promise" object which represents a value and may eventually resolve to either a
+value or object reference or may break, either explicitly, implicitly through
+error handling, or via network partition.
+
+Messages can be sent to a promise as if it were the resolved object. These sent
+messages will be relayed to the eventual object if it is `fulfill`ed to one.
+There is also a "promise resolver" object which is used to provide the promise a
+value or break on an error. The resolver object has two methods:
+
+- `fulfill`: Provide the promise with its resolved values (success case)
+- `break`: Breaks the promise (usually due to an error)
+
+### `fulfill`
+
+This method can take any number of arguments. These arguments can be promise
+pipelined on with the [`op:pick`](#op-pick) operator and should be delivered to
+any listeners of the promise.
+
+### `break`
+
+Break takes a single argument, which represents an error that occured. This
+error should be delivered to any listeners of the promise.
+
+**NOTE:** The value of errors transmitted over CapTP is up to the transmitting
+party. However, including the contents of exception objects or debugging
+information such as backtraces could unintentionally leak sensitive information.
+It may be valuable at a CapTP border to strip certain kinds of debugging
+information, to encrypt its contents, or to use the sealer/unsealer pattern from
+capability literature to secure its contents so that only parties with the
+relevant sealers/unsealers can safely decode them.
+
+## [Promise pipelining](#promise-pipelining)
+
+When an [`op:deliver`](#op-deliver) message is sent and a result is pending, it
+is often desirable to use this result. One way to do this would be to wait until
+the promise has been resolved and then send the next message to the result.
+While this approach can be taken within OCapN, promise pipelining is preferred.
+When a message is sent with an `answer-pos` specified, the promise on the remote
+session can then be referenced and messaged with the
+[`desc:answer`](#desc-answer) descriptor. When the result is ready, messages
+sent to this promise will have them delivered to the result.
+
+If the answer contains multiple values, the [`op:pick`](#op-pick) operator
+MUST be used to choose the value to be pipelined on. This works by exporting the
+individual value in another answer position, which then can be used to message
+the promise. If multiple values are pipelined on, without the pick operation,
+the resulting promise MUST be broken with an error and not fulfilled.
```

Comment:
So there are protocol-level affordances for fulfilling a promise to multiple values (and for automatically breaking messages that do not target a specific member of such a collection)? Interesting.

Reply:
Yeah. If the promise only represents one value you of course don't need to use `op:pick`, it's just for those promises which represent multiple values. It used to be that promises could only represent a single value, and so we often encoded multiple values in lists, which was fine, but it prevented promise pipelining and caused a lot of unpacking in our code. We solved this by adding this operator and having it so promises could represent multiple values.